Mathlib/Data/Set/Basic.lean

```diff
@@ -0,0 +1,2749 @@
+  (not_congr subset_singleton_iff_eq).2 (not_or_of_not hs.ne_empty hs.ne_singleton)
+#align set.nontrivial.not_subset_singleton Set.Nontrivial.not_subset_singleton
+
+theorem nontrivial_univ [Nontrivial α] : (univ : Set α).Nontrivial :=
+  let ⟨x, y, hxy⟩ := exists_pair_ne α
+  ⟨x, mem_univ _, y, mem_univ _, hxy⟩
+#align set.nontrivial_univ Set.nontrivial_univ
+
+theorem nontrivial_of_univ_nontrivial (h : (univ : Set α).Nontrivial) : Nontrivial α :=
+  let ⟨x, _, y, _, hxy⟩ := h
+  ⟨⟨x, y, hxy⟩⟩
+#align set.nontrivial_of_univ_nontrivial Set.nontrivial_of_univ_nontrivial
+
+@[simp]
+theorem nontrivial_univ_iff : (univ : Set α).Nontrivial ↔ Nontrivial α :=
+  ⟨nontrivial_of_univ_nontrivial, fun h => @nontrivial_univ _ h⟩
+#align set.nontrivial_univ_iff Set.nontrivial_univ_iff
+
+theorem nontrivial_of_nontrivial (hs : s.Nontrivial) : Nontrivial α :=
+  let ⟨x, _, y, _, hxy⟩ := hs
+  ⟨⟨x, y, hxy⟩⟩
+#align set.nontrivial_of_nontrivial Set.nontrivial_of_nontrivial
+
+/-- `s`, coerced to a type, is a nontrivial type if and only if `s` is a nontrivial set. -/
+@[simp, norm_cast]
+theorem nontrivial_coe_sort {s : Set α} : Nontrivial s ↔ s.Nontrivial := by
+  simp_rw [← nontrivial_univ_iff, Set.Nontrivial, mem_univ, exists_true_left, SetCoe.exists,
+    Subtype.mk_eq_mk]
+  sorry
+#align set.nontrivial_coe_sort Set.nontrivial_coe_sort
+
+alias nontrivial_coe_sort ↔ _ Nontrivial.coe_sort
+
+/-- A type with a set `s` whose `coe_sort` is a nontrivial type is nontrivial.
+For the corresponding result for `subtype`, see `subtype.nontrivial_iff_exists_ne`. -/
+theorem nontrivial_of_nontrivial_coe (hs : Nontrivial s) : Nontrivial α :=
+  nontrivial_of_nontrivial <| nontrivial_coe_sort.1 hs
+#align set.nontrivial_of_nontrivial_coe Set.nontrivial_of_nontrivial_coe
+
+theorem nontrivial_mono {α : Type _} {s t : Set α} (hst : s ⊆ t) (hs : Nontrivial s) :
+    Nontrivial t :=
+  Nontrivial.coe_sort <| (nontrivial_coe_sort.1 hs).mono hst
+#align set.nontrivial_mono Set.nontrivial_mono
+
+@[simp]
+theorem not_subsingleton_iff : ¬s.Subsingleton ↔ s.Nontrivial := by
+  simp_rw [Set.Subsingleton, Set.Nontrivial, not_forall]; rfl
+#align set.not_subsingleton_iff Set.not_subsingleton_iff
+
+@[simp]
+theorem not_nontrivial_iff : ¬s.Nontrivial ↔ s.Subsingleton :=
+  Iff.not_left not_subsingleton_iff.symm
+#align set.not_nontrivial_iff Set.not_nontrivial_iff
+
+alias not_nontrivial_iff ↔ _ Subsingleton.not_nontrivial
+
+alias not_subsingleton_iff ↔ _ Nontrivial.not_subsingleton
+
+theorem univ_eq_true_false : univ = ({True, False} : Set Prop) :=
+  Eq.symm <| eq_univ_of_forall <| fun x => by classical
+    if h : x then
+      simp [h]
+    else
+      simp [h]
+#align set.univ_eq_true_false Set.univ_eq_true_false
+
+section Preorder
+
+variable [Preorder α] [Preorder β] {f : α → β}
+
+theorem monotoneOn_iff_monotone : MonotoneOn f s ↔
+  (@Monotone _ _ (Subtype.instPreorderSubtype _) _ fun a : s => f a) := by
```

Comment:
Bandaid fix for broken instance search, here and below.

Reply:
I've made `Elem` (previously `type_of_set` or something) reducible, so this instance appears, and restored the original statement. See https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/data.2Eset.2Ebasic/near/315272101